File: backend/app/services/password_reset_service.py
import hashlib
import secrets
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.models.password_reset_token import PasswordResetToken
from app.services.email_service import send_password_reset_email

RESET_TOKEN_EXPIRE_MINUTES = 30


def hash_token(token: str) -> str:
    return hashlib.sha256(token.encode("utf-8")).hexdigest()


def create_password_reset_token(db: Session, user_id) -> str:
    plain = secrets.token_urlsafe(48)
    record = PasswordResetToken(
        user_id=user_id,
        token_hash=hash_token(plain),
        expires_at=datetime.utcnow() + timedelta(minutes=RESET_TOKEN_EXPIRE_MINUTES),
        used=False,
    )
    db.add(record)
    db.commit()
    return plain


def consume_password_reset_token(db: Session, plain_token: str) -> str | None:
    token_hash = hash_token(plain_token)
    record = (
        db.query(PasswordResetToken)
        .filter(
            PasswordResetToken.token_hash == token_hash,
            PasswordResetToken.used.is_(False),
        )
        .first()
    )
    if not record or record.expires_at < datetime.utcnow():
        return None

    record.used = True
    db.commit()
    return str(record.user_id)


def request_password_reset(db: Session, email: str) -> bool:
    from app.models.user import User

    user = db.query(User).filter(User.email == email).first()
    if not user:
        return False

    plain = create_password_reset_token(db, user.id)
    send_password_reset_email(user.email, plain)
    return True

File: backend-node/src/services/passwordResetService.js
import crypto from 'crypto'
import { prisma } from '../db/prisma.js'
import { sendPasswordResetEmail } from './emailService.js'

const RESET_TOKEN_EXPIRE_MINUTES = 10

function hashToken(token) {
  return crypto.createHash('sha256').update(token, 'utf8').digest('hex')
}

async function createPasswordResetToken(userId) {
  const plain = crypto.randomBytes(48).toString('base64url')
  await prisma.passwordResetToken.create({
    data: {
      userId,
      tokenHash: hashToken(plain),
      expiresAt: new Date(Date.now() + RESET_TOKEN_EXPIRE_MINUTES * 60 * 1000),
      used: false,
    },
  })
  return plain
}

export async function consumePasswordResetToken(plainToken) {
  const tokenHash = hashToken(plainToken)
  const record = await prisma.passwordResetToken.findFirst({
    where: { tokenHash, used: false },
  })
  if (!record || record.expiresAt < new Date()) return null

  await prisma.passwordResetToken.update({
    where: { id: record.id },
    data: { used: true },
  })
  return record.userId
}

export async function requestPasswordReset(email) {
  const user = await prisma.user.findUnique({ where: { email } })
  if (!user) return false
  const plain = await createPasswordResetToken(user.id)
  await sendPasswordResetEmail(user.email, plain)
  return true
}
